Mobile/ShopLens/MigrationsConsole/Requirements/SqlRequirements.cs

```diff
@@ -0,0 +1,135 @@
+            cmd.Connection = connection;
+            cmd.CommandType = CommandType.Text;
+            cmd.CommandText = commandText;
+            cmd.ExecuteNonQuery();
+        }
+
+        public void DataAdapterDataTableAndLinq()
+        {
+            // DataAdapter, DataSet
+            Console.WriteLine("Data adapter, data table, and linq:");
+            var users = new List<User>();
+            var sessions = new List<ShoppingSession>();
+            
+            var userAdapter = new SqlDataAdapter("SELECT UserId, Name FROM [Users]", connection);
+            var userSet = new DataSet();
+            
+            userAdapter.Fill(userSet, "Users");
+            userAdapter.Dispose();
+
+            foreach (DataRow row in userSet.Tables[0].Rows)
+            {
+                Console.WriteLine($"Id = {row["UserId"]}, Name = {row["Name"]}");
+                users.Add(new User{Name = (string) row["Name"], UserId = (Guid) row["UserId"]});
+            }
+
+            if (!users.Any())
+            {
+                Console.WriteLine("No users found. Please add some users to the database.");
+                return;
+            }
+            
+            var sessionAdapter = new SqlDataAdapter("SELECT ShoppingSessionId, UserId FROM [ShoppingSessions]", connection);
+            var sessionSet = new DataSet();
+            
+            sessionAdapter.Fill(sessionSet, "ShoppingSessions");
+            sessionAdapter.Dispose();
+            
+            foreach (DataRow row in sessionSet.Tables[0].Rows)
+            {
+                sessions.Add(new ShoppingSession {ShoppingSessionId = (Guid) row["ShoppingSessionId"], UserId = (Guid) row["UserId"]});
+            }
+            
+            // LINQ queries
+            var userNamesAggregated = users.Select(x => x.Name).Aggregate((x, y) => $"{x} and {y}");
+            Console.WriteLine("(.Aggregate)User names aggregated: " + userNamesAggregated);
+
+            var firstTwoUsers = users.Take(3).ToList();
+            Console.WriteLine("User names .Take(3): " + string.Join(", ", firstTwoUsers.Select(x => x.Name)));
+
+            var userNamesGrouped = users.GroupBy(x => x);
+            var mostCommonName = userNamesGrouped.First(x => x.Count() == userNamesGrouped.Max(y => y.Count())).Key;
+            Console.WriteLine($"(.GroupBy) Most common username: {mostCommonName}");
+
+            var usersAndShoppingSessions = sessions
+                .Join(users, x => x.UserId, x => x.UserId, (x, y) => new {x.ShoppingSessionId, x.UserId, y.Name})
+                .Select(x => $"Shopping session of {x.Name}({x.UserId}): \t {x.ShoppingSessionId}");
+
+            Console.WriteLine("(.Join) Users and their shopping sessions:");
+            foreach (var userAndShoppingSession in usersAndShoppingSessions)
+            {
+                Console.WriteLine(userAndShoppingSession);
+            }
+        }
+
+        ~SqlRequirements()
+        {
+            connection.Close();
```

Comment:
The finalizer throws an exception when I tested it.
![image](https://user-images.githubusercontent.com/43152072/49699542-1e538600-fbdb-11e8-974f-94d1ac853ed1.png)

Reply:
Could you perhaps try running it once or twice more? Try commenting out some of the statements like lines 36-40 in Program.cs as well.

It's weird, I got it to crash once, then commented all statements out, it went fine, then uncommented the statements (original code) and now it executes fine again (???) I'm not sure what's going on

Reply:
I commented lines 36-40, like you said, then I commented the finalizer of the Requirements class and I got this error.
![image](https://user-images.githubusercontent.com/43152072/49779367-57424680-fd12-11e8-8a5b-920204228180.png)
This was on a local DB, so I switch over to our Azure DB. Then everything was working with with the lines 36-40 and the finalizer commented out.
![image](https://user-images.githubusercontent.com/43152072/49779456-adaf8500-fd12-11e8-965f-727da2f6f9e1.png)
I uncommented lines 36-40 and on our Azure DB it worker like a charm. 
![image](https://user-images.githubusercontent.com/43152072/49779490-e2234100-fd12-11e8-9da2-c74d995f6fdf.png)
Then I uncommented the finalizer and I got the error again.
![image](https://user-images.githubusercontent.com/43152072/49779506-fbc48880-fd12-11e8-992e-454997e02b2e.png)
I tried it again several times on a local DB with the finalizer commented out and I got the Users error, but no Driver error. 
![image](https://user-images.githubusercontent.com/43152072/49779557-4ba34f80-fd13-11e8-9aed-d3c6266cda4e.png)
Then I tried it several times with the finalizer commented out on our Azure DB and it worked okay.

Reply:
Houston, I think we have found the problem. 
![image](https://user-images.githubusercontent.com/43152072/49779662-df751b80-fd13-11e8-8655-9e319a6d4145.png)

Reply:
Still, I didn't know anything about finalizers, so thanks for showing us that they exist.